File: cane-cpp/cane-core/src/player/pose_modifiers.cpp
#include "state.hpp"
#include "constraint_parameters.hpp"
#include "runtime_data_internal.hpp"

namespace cane::detail {
namespace {
const char* const fields[]{"x", "y", "rotationDegrees", "scaleX", "scaleY", "shearXDegrees", "shearYDegrees"};
using Channels = std::array<std::optional<float>, 7>;
Channels channels(const BoneLocal& v) { return {v.x, v.y, v.rotation_degrees, v.scale_x, v.scale_y, v.shear_x_degrees, v.shear_y_degrees}; }
Channels channels(const BoneLocalPatch& v) { return {v.x, v.y, v.rotation_degrees, v.scale_x, v.scale_y, v.shear_x_degrees, v.shear_y_degrees}; }
Channels channels(const BoneLocalAdditive& v) { return {v.x_delta, v.y_delta, v.rotation_degrees_delta, v.scale_x_delta, v.scale_y_delta, v.shear_x_degrees_delta, v.shear_y_degrees_delta}; }
std::size_t target(const animation::Pose& pose, RuntimeCatalogKind kind, const std::string& id, const char* op, const std::string& field) {
    if (id.empty()) throw Error(ErrorCode::invalid_argument, op, "Modifier target ID must not be empty.", field, id);
    const auto& catalog = RuntimeDataAccess::get(pose.data()).index.catalogs[static_cast<std::size_t>(kind)];
    const auto at = catalog.find(id);
    if (at == catalog.end()) throw Error(ErrorCode::not_found, op, "Modifier target does not exist.", field, id);
    return at->second;
}
void edit_bone(animation::Pose& pose, const std::string& id, const Channels& values, bool additive,
               const char* field, const char* op, const std::string& prefix) {
    const auto index = target(pose, RuntimeCatalogKind::bone, id, op, prefix + ".boneId");
    auto& local = pose.locals[index];
    const std::array<float*, 7> into{&local.x, &local.y, &local.rotation_degrees, &local.scale_x, &local.scale_y, &local.shear_x_degrees, &local.shear_y_degrees};
    for (std::size_t component = 0; component < values.size(); ++component) if (values[component]) {
        const float input = *values[component]; const float value = additive ? *into[component] + input : input;
        if (!std::isfinite(input) || !std::isfinite(value))
            throw Error(ErrorCode::non_finite, op, "Modifier input or arithmetic result is non-finite.", prefix + "." + field + "." + fields[component] + (additive ? "Delta" : ""), id);
        *into[component] = value;
    }
}
}
void apply_pose_modifiers(animation::Pose& pose, const PoseModifiers& modifiers, const char* op) {
    for (std::size_t i = 0; i < modifiers.operations.size(); ++i) {
        const auto& value = modifiers.operations[i]; const auto prefix = "modifiers.operations[" + std::to_string(i) + "]";
        if (const auto* replace = std::get_if<ReplaceBoneLocal>(&value)) edit_bone(pose, replace->bone_id, channels(replace->local), false, "local", op, prefix);
        else if (const auto* patch = std::get_if<PatchBoneLocal>(&value)) edit_bone(pose, patch->bone_id, channels(patch->patch), false, "patch", op, prefix);
        else if (const auto* add = std::get_if<AddBoneLocal>(&value)) edit_bone(pose, add->bone_id, channels(add->delta), true, "delta", op, prefix);
        else if (const auto* constraint = std::get_if<PatchConstraint>(&value)) {
            const auto index = target(pose, RuntimeCatalogKind::constraint, constraint->constraint_id, op, prefix + ".constraintId");
            if (public_constraint_kind(pose.model().constraints[index].kind) != override_kind(constraint->parameters))
                throw Error(ErrorCode::invalid_argument, op, "Constraint modifier kind does not match its target.", prefix + ".parameters.type", constraint->constraint_id);
            try { apply_constraint_override(pose.constraints[index], constraint->parameters, op); }
            catch (const Error& e) { throw Error(e.code, op, e.what(), prefix + "." + e.field.value_or("parameters"), constraint->constraint_id, e.detail); }
        } else throw Error(ErrorCode::invalid_argument, op, "Modifier operation has no active value.", prefix + ".operation");
    }
}
}
namespace cane {
RuntimeFrame RuntimePlayer::apply_with_modifiers(const PoseModifiers& modifiers, const SamplingOptions& options) {
    constexpr auto op = "applyWithModifiers";
    return detail::operation(op, [&] {
        auto& p = impl(); p.require_idle(op); const auto sampling = detail::checked_sampling(options, op);
        auto candidate = std::make_unique<detail::PlayerState>(*p.state); constraints::PhysicsBudget budget;
        p.publish(*candidate, p.evaluate(*candidate, budget, sampling, &modifiers, op), p.next_sequence(op));
        p.state.swap(candidate); return p.frame();
    });
}
RuntimeStep RuntimePlayer::advance_with_modifiers(float delta, const PoseModifiers& modifiers, const SamplingOptions& options) {
    return impl().step(delta, true, options, "advanceWithModifiers", &modifiers);
}
}

File: cane-csharp/cane-core/Runtime/PoseModifiers.cs
using System;
using System.Collections.Generic;
using Cane.Animation;
using Cane.Format;

namespace Cane
{
    public sealed class BoneLocalPatch
    {
        public float? X { get; set; }
        public float? Y { get; set; }
        public float? RotationDegrees { get; set; }
        public float? ScaleX { get; set; }
        public float? ScaleY { get; set; }
        public float? ShearXDegrees { get; set; }
        public float? ShearYDegrees { get; set; }
        internal float?[] Values() => new[] { X, Y, RotationDegrees, ScaleX, ScaleY, ShearXDegrees, ShearYDegrees };
    }
    /// <summary>Ordered, reusable procedural edits. AddBoneLocal interprets every supplied channel as an additive delta.</summary>
    public sealed class PoseModifiers
    {
        internal static readonly string[] Fields = { "x", "y", "rotationDegrees", "scaleX", "scaleY", "shearXDegrees", "shearYDegrees" };
        internal readonly Json Document = new Json(new Dictionary<string, Json> { ["operations"] = new Json(new List<Json>()) });
        public int Count => Document["operations"].Items.Count;
        public void Clear() => Document["operations"].Items.Clear();
        public PoseModifiers ReplaceBoneLocal(string boneId, BoneLocal local)
        { var fields = new Dictionary<string, Json>(); for (int i = 0; i < 7; i++) fields.Add(Fields[i], new Json((double)local.Get(i))); Add("replaceBoneLocal", boneId, "local", new Json(fields)); return this; }
        public PoseModifiers PatchBoneLocal(string boneId, BoneLocalPatch patch) { AddValues("patchBoneLocal", boneId, patch, false); return this; }
        public PoseModifiers AddBoneLocal(string boneId, BoneLocalPatch delta) { AddValues("addBoneLocal", boneId, delta, true); return this; }
        public PoseModifiers PatchConstraint(string constraintId, ConstraintOverride parameters)
        {
            Document["operations"].Items.Add(new Json(new Dictionary<string, Json> { ["operation"] = new Json("patchConstraint"), ["constraintId"] = new Json(constraintId), ["parameters"] = Pose.Clone(parameters.Patch) })); return this;
        }
        private void AddValues(string operation, string boneId, BoneLocalPatch patch, bool additive)
        {
            var fields = new Dictionary<string, Json>(); float?[] values = patch.Values();
            for (int i = 0; i < values.Length; i++) if (values[i].HasValue) fields.Add(Fields[i] + (additive ? "Delta" : ""), new Json((double)values[i]!.Value));
            Add(operation, boneId, additive ? "delta" : "patch", new Json(fields));
        }
        private void Add(string operation, string boneId, string field, Json values)
        { Document["operations"].Items.Add(new Json(new Dictionary<string, Json> { ["operation"] = new Json(operation), ["boneId"] = new Json(boneId), [field] = values })); }
    }
    public sealed class EvaluationStats
    {
        public int AnimationSamples { get; internal set; }
        public int ConstraintGeometrySolves { get; internal set; }
        public int FramesPublished { get; internal set; }
        internal EvaluationStats Copy() => new EvaluationStats { AnimationSamples = AnimationSamples, ConstraintGeometrySolves = ConstraintGeometrySolves, FramesPublished = FramesPublished };
    }
    public sealed partial class RuntimePlayer
    {
        public EvaluationStats LastEvaluationStats => Frame.Stats.Copy();
        public RuntimeFrame ApplyWithModifiers(PoseModifiers modifiers, SamplingOptions sampling = default)
        { if (modifiers == null) throw new RuntimeException(RuntimeErrorCode.InvalidArgument, "applyWithModifiers", "Modifiers are required."); return ApplyWithModifierDocument(modifiers.Document, sampling); }
        internal RuntimeFrame ApplyWithModifierDocument(Json modifiers, SamplingOptions sampling = default)
        {
            const string op = "applyWithModifiers"; sampling.Validate(op);
            try { Playback state = playback.Clone(); RuntimeFrame frame = Evaluate(state, host, NextSequence(op), sampling, out Pose pose, modifiers, op); playback = state; Frame = frame; publishedPose = pose; return frame; }
            catch (RuntimeException e) { throw Reframe(e, op); }
        }
        public RuntimeStep AdvanceWithModifiers(float deltaSeconds, PoseModifiers modifiers, SamplingOptions sampling = default)
        { if (modifiers == null) throw new RuntimeException(RuntimeErrorCode.InvalidArgument, "advanceWithModifiers", "Modifiers are required."); return AdvanceWithModifierDocument(deltaSeconds, modifiers.Document, sampling); }
        internal RuntimeStep AdvanceWithModifierDocument(float delta, Json modifiers, SamplingOptions sampling = default) => AdvanceInternal(delta, true, sampling, modifiers, "advanceWithModifiers");
        private void ApplyTransient(Pose pose, Json? batch, string op)
        {
            if (batch == null) return;
            Validation.Closed(batch, "operations", "operations"); int index = 0;
            foreach (Json operation in batch["operations"].Items)
            {
                string prefix = "modifiers.operations[" + index++ + "]", kind = operation.S("operation");
                if (kind == "patchConstraint")
                {
                    Validation.Closed(operation, "operation constraintId parameters", "operation constraintId parameters");
                    Json patch = NormalizeConstraintPatch(operation.S("constraintId"), operation["parameters"], op, out int constraint);
                    foreach (var p in patch.Members) pose.Constraints[constraint].Members[p.Key] = Pose.Clone(p.Value); continue;
                }
                if (kind != "replaceBoneLocal" && kind != "patchBoneLocal" && kind != "addBoneLocal") throw new RuntimeException(RuntimeErrorCode.InvalidArgument, op, "Unknown modifier operation.", prefix + ".operation");
                bool additive = kind == "addBoneLocal", replace = kind == "replaceBoneLocal"; string field = additive ? "delta" : replace ? "local" : "patch";
                Validation.Closed(operation, "operation boneId " + field, "operation boneId " + field);
                string id = operation.S("boneId"); if (!data.BoneIndex.TryGetValue(id, out int bone)) throw new RuntimeException(RuntimeErrorCode.NotFound, op, "Modifier bone is missing.", prefix + ".boneId", id);
                Json values = operation[field]; string allowed = ""; foreach (string f in PoseModifiers.Fields) allowed += f + (additive ? "Delta " : " ");
                Validation.Closed(values, allowed, replace ? allowed : ""); BoneLocal local = pose.Locals[bone];
                for (int i = 0; i < 7; i++)
                {
                    string key = PoseModifiers.Fields[i] + (additive ? "Delta" : ""); if (values[key].IsNull) { if (replace) throw new RuntimeException(RuntimeErrorCode.InvalidArgument, op, "Replace requires all local channels.", prefix + "." + field + "." + key, id); continue; }
                    float value = values[key].Float; if (!Numeric.Finite(value)) throw new RuntimeException(RuntimeErrorCode.InvalidArgument, op, "Modifier input must be finite.", prefix + "." + field + "." + key, id);
                    if (additive) value = local.Get(i) + value;
                    if (!Numeric.Finite(value)) throw new RuntimeException(RuntimeErrorCode.NonFinite, op, "Modifier arithmetic overflowed.", prefix + "." + field + "." + key, id);
                    local = local.With(i, value);
                }
                pose.Locals[bone] = local;
            }
        }
    }
}
